bctl/agent/plugin/web/actions/webdial/webdial.go

```
@@ -50,42 +51,55 @@ func New(logger *logger.Logger,
 		streamOutputChan: ch,
 		remoteHost:       remoteHost,
 		remotePort:       remotePort,
+		interruptChan:    make(chan bool),
 	}, nil
 }
 
-func (s *WebDial) Closed() bool {
-	return s.closed
+func (w *WebDial) Closed() bool {
+	return w.closed
 }
 
-func (e *WebDial) Receive(action string, actionPayload []byte) (string, []byte, error) {
-	switch WebDialSubAction(action) {
-	case WebDialStart:
+func (w *WebDial) Receive(action string, actionPayload []byte) (string, []byte, error) {
+	var rerr error
+
+	switch bzwebdial.WebDialSubAction(action) {
+	case bzwebdial.WebDialStart:
 		var webDialActionRequest WebDialActionPayload
 		if err := json.Unmarshal(actionPayload, &webDialActionRequest); err != nil {
-			rerr := fmt.Errorf("malformed web dial Action payload %v", actionPayload)
-			e.logger.Error(rerr)
-			return action, []byte{}, rerr
+			rerr := fmt.Errorf("malformed web dial action payload: %s", actionPayload)
+			w.logger.Error(rerr)
+		} else {
+			return w.startDial(webDialActionRequest, action)
 		}
-
-		return e.startDial(webDialActionRequest, action)
-	case WebDialDataIn:
-		// Deserialize the action payload, the only action passed is DataIn
-		var dataIn WebDataInActionPayload
+	case bzwebdial.WebDialInput:
+		var dataIn WebInputActionPayload
 		if err := json.Unmarshal(actionPayload, &dataIn); err != nil {
-			rerr := fmt.Errorf("unable to unmarshal dataIn message: %s", err)
-			e.logger.Error(rerr)
-			return "", []byte{}, rerr
+			rerr := fmt.Errorf("unable to unmarshal web dial input message: %s", err)
+			w.logger.Error(rerr)
+		} else {
+			return w.HandleNewHttpRequest(action, dataIn)
 		}
+	case bzwebdial.WebDialInterrupt:
+		w.interruptChan <- true
+
+		// give our streamoutputchan time to process all the messages we sent while the interrupt was getting here
+		time.Sleep(2 * time.Second)
```

Should we sleep after sending the interrupt ack back to the daemon instead of before? Also maybe it makes sense to add a `processInput` bool on the agent side so we dont keep processing input messages after receiving an interrupt?

---

Stream messages and keysplitting messages are sent independently of each other, so what happens is that the agent has queued up all these stream messages that it's been building since the daemon sent the interrupt, and is sending them one-by-one at the datachannel level. If we send a keypslitting messages, that gets sent immediately and will get to the daemon faster than the agent's datachannel can empty the stream message queue.

So this sleep allows the agent's datachannel to finish processing (hopefully, I acknowledge this isn't a perfect solution) the stream message queue and for the daemon to receive and silently ignore them because the second that daemon gets the ack on the interrupt it shuts everything down which then throws errors on the rest of the stream messages it receives for that action, because the action no longer exists.  This line tries to minimize those errors and why it doesn't need to be a perfect solution because those errors won't break anything.

Sorry that was so long

---

Ok thanks for the explanation i think i understand the issue now. So there are a bunch of pending stream messages in the `streanOutputChan` that have already been generated and are waiting to be sent through the data channel. We sleep to allow this queue to be drained before sending an interrupt ack that would otherwise cause errors on the daemon side when handling those stream messages.

Maybe a cleaner alternative solution is we just kept a map of requestId (string) -> interrupt (bool) in the data channel and we can mark the webdial requestId as interrupted there. Then when processing the `streamOutputChan` messages in the data channel we can check if the request has been interrupted and just drop those messages instead of sending them.

So we would add a check here:

https://github.com/bastionzero/bzero/blob/03e71ec9791f874107591c9c2a3e09f4ecb8c3ca/bctl/agent/datachannel/datachannel.go#L231-L239